### PR TITLE
feat: add WRROC models, validators & unit tests

### PR DESCRIPTION
**Pull request by Karanjot786**
This PR introduces several key changes related to the WRROC models and their corresponding validators. Below is a summary of the changes:

### 1) Addition of WRROC Models:

**TES Models:**
        `Executor:` Represents an executor in the Task Execution Service (TES).
        `TESInputs & TESOutputs:` Models representing input and output files in TES.
        `TESLogs:` Represents logs associated with TES tasks.
        `TESData: `Represents a TES task, including all relevant attributes such as executors, inputs, outputs, logs, etc.

**WES Models:**
        `WESRunLog:` This represents a run log in the Workflow Execution Service (WES).
        `WESOutputs:` Represents output files in WES.
        `WESRequest:` Represents a workflow request in WES.
        `WESData:` Represents a WES run, including request, run log, state, task logs, and outputs.

**WRROC Models:**
        `WRROCInputs & WRROCOutputs:` Models representing inputs and outputs in WRROC.
        `WRROCData:` Represents a general WRROC entity with attributes like id, name, description, instrument, object, result, startTime, and endTime.
        `WRROCDataTES & WRROCDataWES:` Models specifically designed for TES and WES conversion, inheriting from WRROCData.
        `WRROCProcess, WRROCWorkflow, WRROCProvenance:` Hierarchical models representing different profiles (Process Run, Workflow Run, and Provenance Run) in WRROC.

### 2) Validators:

- I added functions to validate WRROC data and determine the specific profile (Process, Workflow, or Provenance).
- We implemented validators to check WRROC data for TES and WES conversions, ensuring the presence of required fields for each conversion type.

### 3) Unit Tests:

- Comprehensive unit tests have been added for the WRROC models and validators to ensure they work as expected.
- Tests include validation for correct data, incorrect data, and missing required fields.

### Note:

Test cases for TES and WES converters are not included in this PR. They will be addressed in a subsequent PR to keep the changes manageable and focused.

## Summary by Sourcery

Add WRROC, TES, and WES models along with corresponding validators and unit tests. Enhance CI workflow to run tests.

New Features:
- Introduce WRROC models including WRROCInputs, WRROCOutputs, WRROCData, WRROCDataTES, WRROCDataWES, WRROCProcess, WRROCWorkflow, and WRROCProvenance.
- Add TES models such as Executor, TESInputs, TESOutputs, TESLogs, and TESData.
- Add WES models including WESRunLog, WESOutputs, WESRequest, and WESData.

Enhancements:
- Implement validators for WRROC data to determine specific profiles (Process, Workflow, or Provenance) and ensure required fields for TES and WES conversions.

CI:
- Enable running tests in the CI workflow by uncommenting the relevant lines in the CI configuration.

Tests:
- Add comprehensive unit tests for WRROC models and validators, covering correct data validation, incorrect data handling, and missing required fields.

### CODE REVIEW

**Comment by Karanjot786** [comment]
Hi @uniqueg and @SalihuDickson,

## Summary

This PR introduces several key improvements and fixes based on recent feedback:

- **Refactor WRROC Models:**
  - Created a base class `WRROCDataBase` to hold common fields and methods.
  - Refactored `WRROCData`, `WRROCDataTES`, and `WRROCDataWES` to inherit from `WRROCDataBase` to reduce code duplication and enhance maintainability.

- **Stricter Validation:**
  - Replaced `extra = "allow"` with `extra = "forbid"` in all WRROC models (`WRROCProcess`, `WRROCWorkflow`, `WRROCProvenance`) to enforce stricter validation.

- **Enhanced Validation and Error Collection:**
  - Modified validation functions to collect and report all errors found during validation, ensuring comprehensive feedback.
  - Simplified field validation by removing manual checks for extra fields and leveraging Pydantic's `extra = "forbid"` configuration.

- **Expanded Unit Tests:**
  - Added test cases for edge scenarios such as empty object or result lists, missing optional fields, and incorrect data types.
  - Implemented tests for invalid URLs and other edge cases.
  - Fixed existing tests by allowing empty lists and adding URL validation in the WES conversion validation.

File: crategen/models.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        state (str): The state of the WES run.
+        run_log (WESRunLog): The log of the WES run.
+        task_logs (Optional[list[WESRunLog]]): The logs of individual tasks within the run.
+        outputs (list[WESOutputs]): The outputs of the WES run.
+    """
+    run_id: str
+    request: WESRequest
+    state: str
+    run_log: WESRunLog
+    task_logs: Optional[list[WESRunLog]] = None
```
Comment (SalihuDickson) [inline crategen/models.py:144]: According to the Ga4GH WES specification this field has been deprecated and replaced with the `tes_logs_url` field. You can still include it to support current and older versions, but perhaps we can consider adding a deprecation warning as well as adding support for `tes_logs_url` feild.

Reference: https://ga4gh.github.io/workflow-execution-service-schemas/docs/#tag/Workflow-Runs/operation/GetRunLog

**Review by SalihuDickson — COMMENTED** [review]
In addition to these comments please not that you are not validating the data in either of the converters. Please take a look at that.

**Review by SalihuDickson — APPROVED** [review]

**Comment by SalihuDickson** [comment]
@Karanjot786, you are currently trying to merge to the models branch which is owned by you (it requires no extra permissions to merge).  If this was by accident please change the base branch.

Alternatively you can merge to the models branch and create a new PR to the main branch.

File: crategen/models.py
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/models.py:1]: I'll review the TES models to ensure they cover all necessary fields for conversion without causing errors due to missing attributes.

File: crategen/models.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    name: str
+    description: Optional[str] = ""
+    executors: list[Executor]
+    inputs: list[TESInputs]
+    outputs: list[TESOutputs]
+    creation_time: str
+    logs: list[TESLogs]
+
+    class Config:
+        extra = "forbid"
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/models.py:75]: I'll update the `TESData` and `WESData` models to allow conversion from complete TES and WES objects without throwing errors. I'll also reconsider the use of `extra = "forbid" `and how it might impact this process.

File: crategen/models.py (hunk 1 above)
Comment (Karanjot786) [reply crategen/models.py:144]: I'll include support for the `tes_logs_url` field in the `WESData` model and add a deprecation warning for the `task_logs` field, following the GA4GH WES specification.

**Comment by Karanjot786** [comment]
@SalihuDickson, I noticed the base branch issue. I’ll merge to the models branch and then create a new PR targeting the main branch.
